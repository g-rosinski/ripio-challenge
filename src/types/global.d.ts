declare global {
    interface Window {
        [name: string]: any
    }
}

export {}