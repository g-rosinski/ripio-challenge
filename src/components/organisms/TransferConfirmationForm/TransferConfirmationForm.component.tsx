import React from "react"
import { useForm } from "../../../hooks/useForm/useForm.hook"
import { Button, Form } from "../../atoms"
import { TextField } from "../../molecules/Fields"
import { TransferConfirmationFormProps, TransferConfirmationFormType } from "./TransferConfirmationForm.types"

export const TransferConfirmationForm: React.FC<TransferConfirmationFormProps> = ({
    initialValues,
    onSubmit,
    validateForm,
    editable = true
}) => {
    const { 
        values, 
        errors, 
        handleOnChangeField, 
        handleOnSubmitForm 
    } = useForm<TransferConfirmationFormType>({
        initialValues,
        validateForm,
        onSubmit
    })


    return (
        <Form onSubmit={handleOnSubmitForm}>
            <TextField 
                name="token"
                label="Token"
                helper="Verificar token para confirmar la transferencia"
                value={values.token}
                error={errors.token}
                disabled={!editable}
                onChange={handleOnChangeField}
            />

            <Button type="submit" disabled={!editable}>Verificar</Button>
        </Form>
    )
}
