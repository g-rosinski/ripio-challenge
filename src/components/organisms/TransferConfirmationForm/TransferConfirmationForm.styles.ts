import styled from 'styled-components'

export const CryptoCurrency = styled.div`
    display: flex;
    justify-content: center;
    width: 100%;
`