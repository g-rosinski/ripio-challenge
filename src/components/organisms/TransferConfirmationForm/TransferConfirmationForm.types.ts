import { UseFormProps } from "../../../hooks/useForm/useForm.hook"

export type TransferConfirmationFormType = {
    token:string
}

export type TransferConfirmationFormProps = UseFormProps<TransferConfirmationFormType> & {
    editable?: boolean,
}