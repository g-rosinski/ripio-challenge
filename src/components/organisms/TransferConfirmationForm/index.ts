export * from './TransferConfirmationForm.component'
export * from './TransferConfirmationForm.types'
export * from './TransferConfirmationForm.styles'