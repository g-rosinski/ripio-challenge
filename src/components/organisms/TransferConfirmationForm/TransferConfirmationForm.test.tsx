import React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { TransferConfirmationForm } from './TransferConfirmationForm.component'
import { TransferConfirmationFormType } from './TransferConfirmationForm.types'

describe('<TransferConfirmationForm/>', () => {

    const initialValuesForm: TransferConfirmationFormType = { token: ''}
    const onSubmitMock = jest.fn()
    const validateForm = jest.fn()
    
    const baseProps = {
        initialValues: initialValuesForm,
        onSubmit: onSubmitMock,
        validateForm: validateForm
    }

    beforeEach(() => {
        jest.clearAllMocks()
        validateForm.mockReturnValue({})
    })

    test('Se reenderice correctamente', () => {
        const {container} = render(<TransferConfirmationForm {...baseProps}/>)
        expect(screen.getByRole('textbox',{name: /token/i})).toHaveValue("")
        expect(container).toMatchSnapshot()
    })

    test('Al hacer click verificar si no tiene errores debe hacer submit', () => {
        render(<TransferConfirmationForm {...baseProps}/>)

        userEvent.click(screen.getByRole('button',{name: /verificar/i}))
        expect(validateForm).toBeCalledWith({token: ''})
        expect(onSubmitMock).toBeCalledWith({token: ''})
    })

    test('Al hacer click verificar si tiene errores no debe hacer submit', () => {
        validateForm.mockReturnValue({token: "Token invalido"})
        render(<TransferConfirmationForm {...baseProps} />)

        userEvent.click(screen.getByRole('button', {name: /verificar/i}))
        expect(validateForm).toBeCalledWith({token: ''})
        expect(screen.getByText("Token invalido")).toBeInTheDocument()
        expect(onSubmitMock).not.toBeCalled()
    })

    test('Debe permitir ingresar valor de token inicial', () => {
        render(<TransferConfirmationForm {...baseProps} initialValues={{token: '111222'}} />)

        expect(screen.getByRole('textbox',{name: /token/i})).toHaveValue('111222')
        userEvent.click(screen.getByRole('button',{name: /verificar/i}))
        expect(validateForm).toBeCalledWith({token: '111222'})
        expect(onSubmitMock).toBeCalledWith({token: '111222'})
    })

    test('Si la prop editable esta en false no debe permitir ingresar un valor y cambiar el estilo del form', () => {
        render(<TransferConfirmationForm {...baseProps} editable={false} />)

        const input = screen.getByRole('textbox',{name: /token/i})
        const button = screen.getByRole('button',{name: /verificar/i})
        expect(input).toHaveValue('')

        userEvent.type(input, '123456')
        expect(input).toHaveValue('')

        userEvent.click(button)
        expect(validateForm).not.toBeCalled()
        expect(onSubmitMock).not.toBeCalled()
    })
})