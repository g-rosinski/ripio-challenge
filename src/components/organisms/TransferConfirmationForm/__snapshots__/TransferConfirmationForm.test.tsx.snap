// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransferConfirmationForm/> Se reenderice correctamente 1`] = `
<div>
  <form
    class="sc-jSUZER fMUTQF"
  >
    <div
      class="sc-gswNZR kGxItT"
    >
      <label
        class="sc-dkrFOg fZIonL"
        for="token"
      >
        Token
      </label>
      <input
        class="sc-gKPRtg crDhfD"
        id="token"
        name="token"
        type="text"
        value=""
      />
      <p
        class="sc-bcXHqe hkdUZn"
      >
        Verificar token para confirmar la transferencia
      </p>
    </div>
    <button
      class="sc-hLBbgP dDanPL"
      type="submit"
    >
      Verificar
    </button>
  </form>
</div>
`;
